### DISCUSSION TITLE
ViewRecord shows up as 404

### DISCUSSION THREAD

**Discussion by kolaente** [filamentphp/filament#6443] [Help]
I created a new resource with a view page, as described [here](https://filamentphp.com/docs/2.x/admin/resources/viewing-records) but when clicking an item it only shows a 404. The entity with that id exists.

My model has a primary key called something else than `id` and the primary key is configured correctly in the model's `$primaryKey` property.

**Comment by kolaente** [comment]
Did some digging and it looks like the primary key is not cast to an int, ignoring eloquents primary key type cast and always using a string.

Effectively, if I change this line here:

https://github.com/filamentphp/filament/blob/b90fec939da29939a6340357e111bf01cbc3ce38/packages/admin/src/Resources/Resource.php#L115

to 

```
->resolveRouteBindingQuery(static::getEloquentQuery(), (int)$key, static::getRecordRouteKeyName())
```

it works.

Should I open an issue?

**Reply by danharrin** [reply]
That would break string primary keys for people who use those. I don't understand how a string would cause that to happen, the SQL being generated is the same, no?

**Reply by kolaente** [reply]
This needs a proper fix, my quick and dirty check was only to see if what I was suspecting was indeed the issue.

The final query generated by filament by default is something like this:

```sql
select * from hs_codes where cn_key = "12345"
```

which does not return any results. Note the `"` around the key. What we want is a query like this:

```sql
select * from hs_codes where cn_key = 12345
```

I was able to work around the problem by overriding the `resolveRouteBindingQuery` method in my model like this:

```php
    public function resolveRouteBindingQuery($query, $value, $field = null)
    {
        if ($field === 'cn_key') {
            $value = (int)$value;
        }

        return $query->where($field ?? $this->getRouteKeyName(), $value);
    }
```

**Reply by danharrin** [reply]
If the fix is inside a Laravel method, maybe this is not an issue with how Filament resolves the record?

**Reply by kolaente** [reply]
I don't understand the way Filament resolves these methods enough to be sure whether this is a problem in Laravel or in Filament. I'd assume Filament would use Eloquent's casting of primary keys.

This works correctly:

```php
Route::get('/test/{code}', function (\App\Models\HsCode $code) {
    return $code;
});
```

That looks like it works correctly for Laravel, but not for Filament.

**Comment by passasooz** [comment]
sorry but I've the same problem and I think that the solution cannot be fix library code... someone can help us?
